File: src/core/extraction/text_extractor.cpp
#include "core/extraction/text_extractor.h"
#include "core/shared/logging.h"

#include <QElapsedTimer>
#include <QFile>
#include <QFileInfo>
#include <QProcess>
#include <QStringConverter>

namespace bs {

namespace {

// 50 MB — files beyond this are too large for full-text indexing
constexpr int64_t kMaxFileSizeBytes = 50 * 1024 * 1024;

const QSet<QString>& officeExtensions()
{
    static const QSet<QString> exts = {
        QStringLiteral("doc"),
        QStringLiteral("docx"),
        QStringLiteral("rtf"),
        QStringLiteral("odt"),
    };
    return exts;
}

} // anonymous namespace

const QSet<QString>& TextExtractor::supportedExtensions()
{
    // Built once, lives for the process lifetime.
    static const QSet<QString> exts = {
        // Plain text
        QStringLiteral("txt"),
        QStringLiteral("text"),
        QStringLiteral("log"),
        QStringLiteral("readme"),
        QStringLiteral("changelog"),
        QStringLiteral("license"),
        QStringLiteral("authors"),
        QStringLiteral("todo"),
        QStringLiteral("notes"),

        // Markup / documentation
        QStringLiteral("md"),
        QStringLiteral("markdown"),
        QStringLiteral("rst"),
        QStringLiteral("adoc"),
        QStringLiteral("asciidoc"),
        QStringLiteral("textile"),
        QStringLiteral("org"),
        QStringLiteral("wiki"),
        QStringLiteral("tex"),
        QStringLiteral("latex"),
        QStringLiteral("bib"),
        QStringLiteral("doc"),
        QStringLiteral("docx"),
        QStringLiteral("rtf"),
        QStringLiteral("odt"),

        // Web / markup
        QStringLiteral("html"),
        QStringLiteral("htm"),
        QStringLiteral("xhtml"),
        QStringLiteral("xml"),
        QStringLiteral("xsl"),
        QStringLiteral("xslt"),
        QStringLiteral("svg"),
        QStringLiteral("css"),
        QStringLiteral("scss"),
        QStringLiteral("sass"),
        QStringLiteral("less"),
        QStringLiteral("styl"),

        // JavaScript / TypeScript
        QStringLiteral("js"),
        QStringLiteral("jsx"),
        QStringLiteral("ts"),
        QStringLiteral("tsx"),
        QStringLiteral("mjs"),
        QStringLiteral("cjs"),
        QStringLiteral("vue"),
        QStringLiteral("svelte"),

        // C / C++
        QStringLiteral("c"),
        QStringLiteral("h"),
        QStringLiteral("cpp"),
        QStringLiteral("cxx"),
        QStringLiteral("cc"),
        QStringLiteral("hpp"),
        QStringLiteral("hxx"),
        QStringLiteral("hh"),
        QStringLiteral("ipp"),
        QStringLiteral("inl"),

        // C# / .NET
        QStringLiteral("cs"),
        QStringLiteral("csx"),
        QStringLiteral("fs"),
        QStringLiteral("fsx"),
        QStringLiteral("fsi"),
        QStringLiteral("vb"),

        // Java / JVM
        QStringLiteral("java"),
        QStringLiteral("kt"),
        QStringLiteral("kts"),
        QStringLiteral("scala"),
        QStringLiteral("sc"),
        QStringLiteral("groovy"),
        QStringLiteral("gradle"),
        QStringLiteral("clj"),
        QStringLiteral("cljs"),
        QStringLiteral("cljc"),
        QStringLiteral("edn"),

        // Systems languages
        QStringLiteral("rs"),
        QStringLiteral("go"),
        QStringLiteral("swift"),
        QStringLiteral("m"),
        QStringLiteral("mm"),
        QStringLiteral("zig"),
        QStringLiteral("nim"),
        QStringLiteral("d"),
        QStringLiteral("v"),

        // Scripting
        QStringLiteral("py"),
        QStringLiteral("pyi"),
        QStringLiteral("pyw"),
        QStringLiteral("rb"),
        QStringLiteral("rbw"),
        QStringLiteral("pl"),
        QStringLiteral("pm"),
        QStringLiteral("t"),
        QStringLiteral("php"),
        QStringLiteral("phps"),
        QStringLiteral("lua"),
        QStringLiteral("tcl"),
        QStringLiteral("r"),
        QStringLiteral("rmd"),
        QStringLiteral("jl"),
        QStringLiteral("ex"),
        QStringLiteral("exs"),
        QStringLiteral("erl"),
        QStringLiteral("hrl"),
        QStringLiteral("hs"),
        QStringLiteral("lhs"),
        QStringLiteral("ml"),
        QStringLiteral("mli"),
        QStringLiteral("sml"),

        // Shell
        QStringLiteral("sh"),
        QStringLiteral("bash"),
        QStringLiteral("zsh"),
        QStringLiteral("fish"),
        QStringLiteral("csh"),
        QStringLiteral("ksh"),
        QStringLiteral("bat"),
        QStringLiteral("cmd"),
        QStringLiteral("ps1"),
        QStringLiteral("psm1"),
        QStringLiteral("psd1"),

        // Data / config
        QStringLiteral("json"),
        QStringLiteral("jsonl"),
        QStringLiteral("jsonc"),
        QStringLiteral("json5"),
        QStringLiteral("yaml"),
        QStringLiteral("yml"),
        QStringLiteral("toml"),
        QStringLiteral("ini"),
        QStringLiteral("cfg"),
        QStringLiteral("conf"),
        QStringLiteral("config"),
        QStringLiteral("properties"),
        QStringLiteral("env"),
        QStringLiteral("csv"),
        QStringLiteral("tsv"),
        QStringLiteral("sql"),
        QStringLiteral("graphql"),
        QStringLiteral("gql"),
        QStringLiteral("proto"),
        QStringLiteral("thrift"),
        QStringLiteral("avsc"),

        // Build / CI
        QStringLiteral("cmake"),
        QStringLiteral("make"),
        QStringLiteral("makefile"),
        QStringLiteral("mk"),
        QStringLiteral("dockerfile"),
        QStringLiteral("vagrantfile"),
        QStringLiteral("rakefile"),
        QStringLiteral("gemfile"),
        QStringLiteral("podfile"),
        QStringLiteral("fastfile"),

        // Misc
        QStringLiteral("diff"),
        QStringLiteral("patch"),
        QStringLiteral("gitignore"),
        QStringLiteral("gitattributes"),
        QStringLiteral("gitmodules"),
        QStringLiteral("editorconfig"),
        QStringLiteral("htaccess"),
        QStringLiteral("nginx"),
        QStringLiteral("tf"),
        QStringLiteral("tfvars"),
        QStringLiteral("hcl"),
        QStringLiteral("plist"),
        QStringLiteral("pbxproj"),
    };
    return exts;
}

bool TextExtractor::supports(const QString& extension) const
{
    return supportedExtensions().contains(extension.toLower());
}

ExtractionResult TextExtractor::extract(const QString& filePath)
{
    QElapsedTimer timer;
    timer.start();

    ExtractionResult result;

    // Check file accessibility and size
    QFileInfo info(filePath);
    if (!info.exists() || !info.isFile()) {
        result.status = ExtractionResult::Status::Inaccessible;
        result.errorMessage = QStringLiteral("File does not exist or is not a regular file");
        result.durationMs = static_cast<int>(timer.elapsed());
        return result;
    }

    if (!info.isReadable()) {
        result.status = ExtractionResult::Status::Inaccessible;
        result.errorMessage = QStringLiteral("File is not readable");
        result.durationMs = static_cast<int>(timer.elapsed());
        return result;
    }

    if (info.size() > kMaxFileSizeBytes) {
        result.status = ExtractionResult::Status::SizeExceeded;
        result.errorMessage = QString("File size %1 bytes exceeds limit of %2 bytes")
                                  .arg(info.size())
                                  .arg(kMaxFileSizeBytes);
        result.durationMs = static_cast<int>(timer.elapsed());
        LOG_INFO(bsExtraction, "Skipping oversized file: %s (%lld bytes)",
                 qUtf8Printable(filePath), static_cast<long long>(info.size()));
        return result;
    }

    const QString extension = info.suffix().toLower();
    if (officeExtensions().contains(extension)) {
        // macOS-native Office/RTF/ODT conversion path.
        QProcess textutil;
        QStringList args = {
            QStringLiteral("-convert"),
            QStringLiteral("txt"),
            QStringLiteral("-stdout"),
            filePath,
        };
        textutil.start(QStringLiteral("/usr/bin/textutil"), args);

        if (!textutil.waitForFinished(30000)) {
            textutil.kill();
            textutil.waitForFinished();
            result.status = ExtractionResult::Status::Timeout;
            result.errorMessage = QStringLiteral("textutil conversion timed out");
            result.durationMs = static_cast<int>(timer.elapsed());
            return result;
        }

        const int exitCode = textutil.exitCode();
        const bool ok = (textutil.exitStatus() == QProcess::NormalExit && exitCode == 0);
        if (!ok) {
            const QString stderrText = QString::fromUtf8(textutil.readAllStandardError()).trimmed();
            result.status = ExtractionResult::Status::UnsupportedFormat;
            result.errorMessage = stderrText.isEmpty()
                                      ? QStringLiteral("textutil conversion failed")
                                      : QStringLiteral("textutil conversion failed: %1")
                                            .arg(stderrText.left(200));
            result.durationMs = static_cast<int>(timer.elapsed());
            return result;
        }

        QByteArray rawBytes = textutil.readAllStandardOutput();
        QString decoded = QString::fromUtf8(rawBytes);
        if (decoded.isNull()) {
            decoded = QString::fromLatin1(rawBytes);
        }

        result.status = ExtractionResult::Status::Success;
        result.content = std::move(decoded);
        result.durationMs = static_cast<int>(timer.elapsed());
        return result;
    }

    // Open and read file contents
    QFile file(filePath);
    if (!file.open(QIODevice::ReadOnly)) {
        result.status = ExtractionResult::Status::Inaccessible;
        result.errorMessage = QString("Failed to open file: %1").arg(file.errorString());
        result.durationMs = static_cast<int>(timer.elapsed());
        return result;
    }

    const QByteArray rawBytes = file.readAll();
    file.close();

    if (rawBytes.isEmpty()) {
        // Empty file is valid — just has no content
        result.status = ExtractionResult::Status::Success;
        result.content = QString();
        result.durationMs = static_cast<int>(timer.elapsed());
        return result;
    }

    // Attempt UTF-8 decoding first
    QString decoded;
    {
        auto toUtf8 = QStringDecoder(QStringDecoder::Utf8,
                                     QStringDecoder::Flag::Stateless);
        decoded = toUtf8(rawBytes);

        if (toUtf8.hasError()) {
            // UTF-8 failed — fall back to Latin-1 (always succeeds, byte-for-byte)
            decoded = QString::fromLatin1(rawBytes);
            LOG_DEBUG(bsExtraction, "UTF-8 decode failed for %s, using Latin-1 fallback",
                      qUtf8Printable(filePath));
        }
    }

    result.status = ExtractionResult::Status::Success;
    result.content = std::move(decoded);
    result.durationMs = static_cast<int>(timer.elapsed());

    LOG_DEBUG(bsExtraction, "Extracted %lld chars from %s in %d ms",
              static_cast<long long>(result.content->size()),
              qUtf8Printable(filePath),
              result.durationMs);

    return result;
}

} // namespace bs

File: Core/Extraction/TextExtractors/TextExtractor.swift
// DEPRECATED SWIFT REFERENCE
// Qt/C++ is the source of truth.
// Keep this file only as temporary migration reference while parity items are closed.
// Do not add new features or fixes here.

import Foundation

/// Protocol for text extraction from files
public protocol TextExtractor: Sendable {
    /// File extensions this extractor handles
    var supportedExtensions: Set<String> { get }

    /// Extract text content from a file
    func extract(from url: URL) async throws -> ExtractionResult
}

/// Result of text extraction
public struct ExtractionResult: Sendable {
    public let text: String
    public let metadata: ExtractionMetadata
    public let chunks: [TextChunk]

    public init(text: String, metadata: ExtractionMetadata, chunks: [TextChunk]) {
        self.text = text
        self.metadata = metadata
        self.chunks = chunks
    }
}

/// Metadata extracted from a file
public struct ExtractionMetadata: Sendable {
    public let title: String?
    public let author: String?
    public let createdDate: Date?
    public let modifiedDate: Date?
    public let wordCount: Int
    public let characterCount: Int
    public let language: String?
    public let encoding: String?
    public let mimeType: String?

    public init(
        title: String? = nil,
        author: String? = nil,
        createdDate: Date? = nil,
        modifiedDate: Date? = nil,
        wordCount: Int = 0,
        characterCount: Int = 0,
        language: String? = nil,
        encoding: String? = nil,
        mimeType: String? = nil
    ) {
        self.title = title
        self.author = author
        self.createdDate = createdDate
        self.modifiedDate = modifiedDate
        self.wordCount = wordCount
        self.characterCount = characterCount
        self.language = language
        self.encoding = encoding
        self.mimeType = mimeType
    }
}

/// A chunk of extracted text for indexing
public struct TextChunk: Sendable {
    public let index: Int
    public let text: String
    public let startOffset: Int
    public let endOffset: Int

    public init(index: Int, text: String, startOffset: Int, endOffset: Int) {
        self.index = index
        self.text = text
        self.startOffset = startOffset
        self.endOffset = endOffset
    }
}

/// Errors from extraction
public enum ExtractionError: Error, LocalizedError {
    case unsupportedFileType(String)
    case fileNotReadable(String)
    case encodingError(String)
    case extractionFailed(String)
    case fileTooLarge(Int64)

    public var errorDescription: String? {
        switch self {
        case .unsupportedFileType(let ext):
            return "Unsupported file type: \(ext)"
        case .fileNotReadable(let path):
            return "File not readable: \(path)"
        case .encodingError(let message):
            return "Encoding error: \(message)"
        case .extractionFailed(let message):
            return "Extraction failed: \(message)"
        case .fileTooLarge(let size):
            return "File too large: \(ByteCountFormatter.string(fromByteCount: size, countStyle: .file))"
        }
    }
}

/// Plain text extractor for txt, md, code files
public struct PlainTextExtractor: TextExtractor {
    public let supportedExtensions: Set<String> = [
        // Plain text
        "txt", "text", "md", "markdown", "rst", "asciidoc", "adoc",
        // Code files
        "swift", "m", "h", "c", "cpp", "cc", "cxx", "hpp", "hxx",
        "py", "pyw", "pyx", "pxd", "pxi",
        "js", "jsx", "ts", "tsx", "mjs", "cjs",
        "java", "kt", "kts", "scala", "groovy", "gradle",
        "rb", "rake", "gemspec",
        "go", "rs", "zig",
        "php", "phtml",
        "cs", "fs", "fsx", "vb",
        "lua", "r", "R", "jl", "pl", "pm",
        "sh", "bash", "zsh", "fish", "ps1", "psm1",
        "sql", "graphql", "gql",
        "html", "htm", "xhtml", "xml", "xsl", "xslt",
        "css", "scss", "sass", "less", "styl",
        "json", "jsonc", "json5", "yaml", "yml", "toml", "ini", "cfg", "conf",
        "dockerfile", "containerfile",
        "makefile", "cmake",
        "vim", "vimrc", "el", "lisp", "clj", "cljs", "edn",
        "erl", "hrl", "ex", "exs",
        "hs", "lhs", "ml", "mli", "mll", "mly",
        "tf", "tfvars", "hcl",
        "proto", "thrift", "avsc",
        "vue", "svelte", "astro",
        // Config files (no extension but common names)
        "gitignore", "gitattributes", "editorconfig", "prettierrc", "eslintrc",
    ]

    private let maxFileSize: Int64
    private let chunkSize: Int

    public init(maxFileSize: Int64 = 50 * 1024 * 1024, chunkSize: Int = 1000) {
        self.maxFileSize = maxFileSize
        self.chunkSize = chunkSize
    }

    public func extract(from url: URL) async throws -> ExtractionResult {
        // Check file size
        let attrs = try FileManager.default.attributesOfItem(atPath: url.path)
        let fileSize = (attrs[.size] as? Int64) ?? 0

        if fileSize > maxFileSize {
            throw ExtractionError.fileTooLarge(fileSize)
        }

        // Try to read with detected encoding
        let text: String
        do {
            text = try String(contentsOf: url, encoding: .utf8)
        } catch {
            // Try other encodings
            if let contents = try? String(contentsOf: url, encoding: .isoLatin1) {
                text = contents
            } else if let contents = try? String(contentsOf: url, encoding: .macOSRoman) {
                text = contents
            } else {
                throw ExtractionError.encodingError("Could not decode file with any supported encoding")
            }
        }

        let chunks = chunkText(text)

        let metadata = ExtractionMetadata(
            wordCount: countWords(in: text),
            characterCount: text.count,
            encoding: "utf-8",
            mimeType: mimeType(for: url.pathExtension)
        )

        return ExtractionResult(text: text, metadata: metadata, chunks: chunks)
    }

    private func chunkText(_ text: String) -> [TextChunk] {
        var chunks: [TextChunk] = []
        var currentIndex = text.startIndex
        var chunkIndex = 0
        var offset = 0

        while currentIndex < text.endIndex {
            let remainingDistance = text.distance(from: currentIndex, to: text.endIndex)
            let chunkLength = min(chunkSize, remainingDistance)
            let endIndex = text.index(currentIndex, offsetBy: chunkLength)

            // Try to break at a word boundary
            var actualEnd = endIndex
            if endIndex < text.endIndex {
                if let spaceIndex = text[currentIndex..<endIndex].lastIndex(of: " ") {
                    actualEnd = text.index(after: spaceIndex)
                } else if let newlineIndex = text[currentIndex..<endIndex].lastIndex(of: "\n") {
                    actualEnd = text.index(after: newlineIndex)
                }
            }

            let chunkText = String(text[currentIndex..<actualEnd])
            let chunkEndOffset = offset + text.distance(from: currentIndex, to: actualEnd)

            chunks.append(TextChunk(
                index: chunkIndex,
                text: chunkText,
                startOffset: offset,
                endOffset: chunkEndOffset
            ))

            offset = chunkEndOffset
            currentIndex = actualEnd
            chunkIndex += 1
        }

        return chunks
    }

    private func countWords(in text: String) -> Int {
        let words = text.components(separatedBy: .whitespacesAndNewlines)
            .filter { !$0.isEmpty }
        return words.count
    }

    private func mimeType(for ext: String) -> String {
        switch ext.lowercased() {
        case "txt", "text": return "text/plain"
        case "md", "markdown": return "text/markdown"
        case "html", "htm": return "text/html"
        case "css": return "text/css"
        case "js", "mjs": return "text/javascript"
        case "json": return "application/json"
        case "xml": return "application/xml"
        case "yaml", "yml": return "application/x-yaml"
        default: return "text/plain"
        }
    }
}
